arangod/RestHandler/RestReplicationHandler.cpp

```diff
@@ -2700,16 +2589,144 @@ ReplicationApplier* RestReplicationHandler::getApplier(bool& global) {
+    aql::QueryPart::PART_MAIN /* Do locking */
+  );
+ // NOTE: The collections are on purpose not locked here.
+  // To accuire an EXCLUSIVE lock may require time under load,
+  // we want to allow to cancel this operation while waiting
+  // for the lock.
 
-arangodb::basics::ConditionVariable RestReplicationHandler::_condVar;
+  auto queryRegistry = QueryRegistryFeature::QUERY_REGISTRY;
+  TRI_ASSERT(queryRegistry != nullptr);
+  if (queryRegistry == nullptr) {
+    return {TRI_ERROR_SHUTTING_DOWN};
+  }
 
-//////////////////////////////////////////////////////////////////////////////
-/// @brief global table of flags to cancel holdReadLockCollection jobs, if
-/// the flag is set of the ID of a job, the job is cancelled
-//////////////////////////////////////////////////////////////////////////////
+  {
+    // we need to lock in EXCLUSIVE mode here, because simply locking
+    // in READ mode will not stop other writers in RocksDB. In order
+    // to stop other writers, we need to fetch the EXCLUSIVE lock
+    AccessMode::Type access = EngineSelectorFeature::ENGINE->typeName() == "rocksdb"
+      ? AccessMode::Type::EXCLUSIVE : AccessMode::Type::READ;
+    auto ctx = transaction::StandaloneContext::Create(_vocbase);
+    auto trx = std::make_unique<SingleCollectionTransaction>(ctx, col, access);
+    query->setTransactionContext(ctx);
+    query->injectTransaction(trx.release());
+  }
+  auto trx = query->trx();
+  TRI_ASSERT(trx != nullptr);
+  trx->addHint(transaction::Hints::Hint::LOCK_ENTIRELY);
+ 
+  TRI_ASSERT(isLockHeld(id).is(TRI_ERROR_HTTP_NOT_FOUND));
+
+  try {
+    queryRegistry->insert(id, query.get(), ttl, true, true);
+  } catch (...) {
+    // For compatibility we only return this error
+    return {TRI_ERROR_TRANSACTION_INTERNAL, "cannot begin read transaction"};
+  }
+  // For leak protection in case of errors the unique_ptr
+  // is not responsible anymore, it has been handed over to the
+  // registry.
+  auto q = query.release();
+  // Make sure to return the query after we are done
+  TRI_DEFER(queryRegistry->close(&_vocbase, id));
+
+  TRI_ASSERT(isLockHeld(id).ok());
+  TRI_ASSERT(isLockHeld(id).get() == false);
+
+  TRI_IF_FAILURE("RestReplicationHandler::destroyBeforeLock") {
+    // This will flag the query as destroyed while getting the
+    // blocking lock
+    cancelBlockingTransaction(id);
+  }
+
+  return q->trx()->begin();
+}
 
-std::unordered_map<std::string, std::shared_ptr<SingleCollectionTransaction>>
-    RestReplicationHandler::_holdReadLockJobs;
+ResultT<bool> RestReplicationHandler::isLockHeld(aql::QueryId id) const {
+  // We try to get a lease of the query.
+  // If this is sucessful we will have the lock,
+  // otherwise we do not have it.
+  // The query is only hold for long during initial locking
+  // there it should return false.
+  // In all other cases it is released quickly.
+  auto queryRegistry = QueryRegistryFeature::QUERY_REGISTRY;
+  TRI_ASSERT(queryRegistry != nullptr);
```

Comment:
can we remove these asserts if we cannot guarantee that there will be no shutdown?